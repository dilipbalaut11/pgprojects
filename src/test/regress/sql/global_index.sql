--
-- GLOBAL index tests
--
CREATE TABLE range_parted (
	a int,
	b int
) PARTITION BY RANGE (a);

--create some partitions and insert data
CREATE TABLE range_parted_1 PARTITION OF range_parted FOR VALUES FROM (1) TO (100000);
CREATE TABLE range_parted_2 PARTITION OF range_parted FOR VALUES FROM (100000) TO (200000);
CREATE TABLE range_parted_3 PARTITION OF range_parted FOR VALUES FROM (200000) TO (300000);
INSERT INTO range_parted SELECT i,i%100 FROM generate_series(1,299999) AS i;

--Create global index
CREATE INDEX global_idx ON range_parted(b) global;
INSERT INTO range_parted SELECT i,i%200 FROM generate_series(1,299999) AS i;
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 7;
SELECT * FROM range_parted WHERE b = 7 LIMIT 10;

EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 110;
SELECT * FROM range_parted WHERE b = 110 LIMIT 10;
SELECT * FROM range_parted WHERE b = 250 LIMIT 10;

UPDATE range_parted SET b=b+100;
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 250;
SELECT * FROM range_parted WHERE b = 250 LIMIT 10;

--attach partition
CREATE TABLE range_parted_4(a int, b int);
INSERT INTO range_parted_4 SELECT i,i%300 + 100 FROM generate_series(300000,300300) as i;
ALTER TABLE range_parted ATTACH PARTITION range_parted_4 FOR VALUES FROM (300000) to (400000);
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 350;
SELECT * FROM range_parted WHERE b = 350 LIMIT 10;

CREATE TABLE range_parted_5(a int, b int) PARTITION BY RANGE (a);
CREATE TABLE range_parted_5_1 PARTITION OF range_parted_5 FOR VALUES FROM (400000) TO (450000);
CREATE TABLE range_parted_5_2 PARTITION OF range_parted_5 FOR VALUES FROM (450000) TO (460000);
INSERT INTO range_parted_5 SELECT i,i%100 + 500 FROM generate_series(400000,459999) AS i;
CREATE INDEX global_idx_1 ON range_parted_5(b) global;
EXPLAIN (COSTS OFF) SELECT * FROM range_parted_5 WHERE b = 550;
SELECT * FROM range_parted_5 WHERE b = 550 LIMIT 5;

--attach to the top partition
ALTER TABLE range_parted ATTACH PARTITION range_parted_5 FOR VALUES FROM (400000) to (500000);
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 550;
SELECT * FROM range_parted WHERE b = 550 LIMIT 5;

-- Attach to level-1 partition (test with multi level global index)
CREATE TABLE range_parted_6(a int, b int);
INSERT INTO range_parted_6 SELECT i,i%100 + 600 FROM generate_series(460000,490000) AS i;
ALTER TABLE range_parted_5 ATTACH PARTITION range_parted_6 FOR VALUES FROM (460000) to (500000);
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 650;
SELECT * FROM range_parted WHERE b = 650 LIMIT 5;

-- Update the leaf and check we are inserting that in multi-level global index
UPDATE range_parted SET b=b+1000;
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 1650;
SELECT * FROM range_parted WHERE b = 1650 LIMIT 5;
EXPLAIN (COSTS OFF) SELECT * FROM range_parted_5 WHERE b = 1650;
SELECT * FROM range_parted_5 WHERE b = 1650 LIMIT 5;

-- Conditional update using global index
EXPLAIN (COSTS OFF) UPDATE range_parted SET b=b+1000 where b = 1650;
UPDATE range_parted SET b=b+1000 where b = 1650;
EXPLAIN (COSTS OFF) SELECT * FROM range_parted_5 WHERE b = 2650;
SELECT * FROM range_parted_5 WHERE b = 2650 LIMIT 5;

--Detach partition
ALTER TABLE range_parted DETACH PARTITION range_parted_5;
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 550;
SELECT * FROM range_parted WHERE b = 550 LIMIT 5;

--Reattach the partition
ALTER TABLE range_parted ATTACH PARTITION range_parted_5 FOR VALUES FROM (400000) to (500000);
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 550;
SELECT * FROM range_parted WHERE b = 550 LIMIT 5;

--Drop the partitioned table
DROP TABLE range_parted_5;
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 550;
SELECT * FROM range_parted WHERE b = 550 LIMIT 5;

-- Test unique global index
TRUNCATE TABLE range_parted;
INSERT INTO range_parted VALUES(1,2);
INSERT INTO range_parted VALUES(2,2);
CREATE UNIQUE INDEX global_idx_unique ON range_parted(b) global; -- Fail with duplicate
TRUNCATE TABLE range_parted;
INSERT INTO range_parted VALUES(1,2);
CREATE UNIQUE INDEX global_idx_unique ON range_parted(b) global;
INSERT INTO range_parted VALUES(1,2); -- Fail with duplicate
DROP INDEX global_idx_unique;
INSERT INTO range_parted VALUES(1,2); -- Now this should pass
-- Cleanup
DROP TABLE range_parted;
