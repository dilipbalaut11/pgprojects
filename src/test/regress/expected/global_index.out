--
-- GLOBAL index tests
--
CREATE TABLE range_parted (
	a int,
	b int
) PARTITION BY RANGE (a);
--create some partitions and insert data
CREATE TABLE range_parted_1 PARTITION OF range_parted FOR VALUES FROM (1) TO (100000);
CREATE TABLE range_parted_2 PARTITION OF range_parted FOR VALUES FROM (100000) TO (200000);
CREATE TABLE range_parted_3 PARTITION OF range_parted FOR VALUES FROM (200000) TO (300000);
INSERT INTO range_parted SELECT i,i%100 FROM generate_series(1,299999) AS i;
--Create global index
CREATE INDEX global_idx ON range_parted(b) global;
INSERT INTO range_parted SELECT i,i%200 FROM generate_series(1,299999) AS i;
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 7;
                     QUERY PLAN                     
----------------------------------------------------
 Global Index Scan using global_idx on range_parted
   Index Cond: (b = 7)
(2 rows)

SELECT * FROM range_parted WHERE b = 7 LIMIT 10;
  a  | b 
-----+---
   7 | 7
 107 | 7
 207 | 7
 307 | 7
 407 | 7
 507 | 7
 607 | 7
 707 | 7
 807 | 7
 907 | 7
(10 rows)

EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 110;
                     QUERY PLAN                     
----------------------------------------------------
 Global Index Scan using global_idx on range_parted
   Index Cond: (b = 110)
(2 rows)

SELECT * FROM range_parted WHERE b = 110 LIMIT 10;
  a   |  b  
------+-----
  110 | 110
  310 | 110
  510 | 110
  710 | 110
  910 | 110
 1110 | 110
 1310 | 110
 1510 | 110
 1710 | 110
 1910 | 110
(10 rows)

SELECT * FROM range_parted WHERE b = 250 LIMIT 10;
 a | b 
---+---
(0 rows)

UPDATE range_parted SET b=b+100;
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 250;
                     QUERY PLAN                     
----------------------------------------------------
 Global Index Scan using global_idx on range_parted
   Index Cond: (b = 250)
(2 rows)

SELECT * FROM range_parted WHERE b = 250 LIMIT 10;
  a   |  b  
------+-----
  150 | 250
  350 | 250
  550 | 250
  750 | 250
  950 | 250
 1150 | 250
 1350 | 250
 1550 | 250
 1750 | 250
 1950 | 250
(10 rows)

--attach partition
CREATE TABLE range_parted_4(a int, b int);
INSERT INTO range_parted_4 SELECT i,i%300 + 100 FROM generate_series(300000,300300) as i;
ALTER TABLE range_parted ATTACH PARTITION range_parted_4 FOR VALUES FROM (300000) to (400000);
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 350;
                     QUERY PLAN                     
----------------------------------------------------
 Global Index Scan using global_idx on range_parted
   Index Cond: (b = 350)
(2 rows)

SELECT * FROM range_parted WHERE b = 350 LIMIT 10;
   a    |  b  
--------+-----
 300250 | 350
(1 row)

CREATE TABLE range_parted_5(a int, b int) PARTITION BY RANGE (a);
CREATE TABLE range_parted_5_1 PARTITION OF range_parted_5 FOR VALUES FROM (400000) TO (450000);
CREATE TABLE range_parted_5_2 PARTITION OF range_parted_5 FOR VALUES FROM (450000) TO (460000);
INSERT INTO range_parted_5 SELECT i,i%100 + 500 FROM generate_series(400000,459999) AS i;
CREATE INDEX global_idx_1 ON range_parted_5(b) global;
EXPLAIN (COSTS OFF) SELECT * FROM range_parted_5 WHERE b = 550;
                       QUERY PLAN                       
--------------------------------------------------------
 Global Index Scan using global_idx_1 on range_parted_5
   Index Cond: (b = 550)
(2 rows)

SELECT * FROM range_parted_5 WHERE b = 550 LIMIT 5;
   a    |  b  
--------+-----
 400050 | 550
 400150 | 550
 400250 | 550
 400350 | 550
 400450 | 550
(5 rows)

--attach to the top partition
ALTER TABLE range_parted ATTACH PARTITION range_parted_5 FOR VALUES FROM (400000) to (500000);
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 550;
                     QUERY PLAN                     
----------------------------------------------------
 Global Index Scan using global_idx on range_parted
   Index Cond: (b = 550)
(2 rows)

SELECT * FROM range_parted WHERE b = 550 LIMIT 5;
   a    |  b  
--------+-----
 400050 | 550
 400150 | 550
 400250 | 550
 400350 | 550
 400450 | 550
(5 rows)

-- Attach to level-1 partition (test with multi level global index)
CREATE TABLE range_parted_6(a int, b int);
INSERT INTO range_parted_6 SELECT i,i%100 + 600 FROM generate_series(460000,490000) AS i;
ALTER TABLE range_parted_5 ATTACH PARTITION range_parted_6 FOR VALUES FROM (460000) to (500000);
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 650;
                     QUERY PLAN                     
----------------------------------------------------
 Global Index Scan using global_idx on range_parted
   Index Cond: (b = 650)
(2 rows)

SELECT * FROM range_parted WHERE b = 650 LIMIT 5;
   a    |  b  
--------+-----
 460050 | 650
 460150 | 650
 460250 | 650
 460350 | 650
 460450 | 650
(5 rows)

-- Update the leaf and check we are inserting that in multi-level global index
UPDATE range_parted SET b=b+1000;
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 1650;
                     QUERY PLAN                     
----------------------------------------------------
 Global Index Scan using global_idx on range_parted
   Index Cond: (b = 1650)
(2 rows)

SELECT * FROM range_parted WHERE b = 1650 LIMIT 5;
   a    |  b   
--------+------
 460050 | 1650
 460150 | 1650
 460250 | 1650
 460350 | 1650
 460450 | 1650
(5 rows)

EXPLAIN (COSTS OFF) SELECT * FROM range_parted_5 WHERE b = 1650;
                       QUERY PLAN                       
--------------------------------------------------------
 Global Index Scan using global_idx_1 on range_parted_5
   Index Cond: (b = 1650)
(2 rows)

SELECT * FROM range_parted_5 WHERE b = 1650 LIMIT 5;
   a    |  b   
--------+------
 460050 | 1650
 460150 | 1650
 460250 | 1650
 460350 | 1650
 460450 | 1650
(5 rows)

-- conditional update using global index -- currently not working
--UPDATE range_parted SET b=b+1000 where b = 1650;
--Detach partition
ALTER TABLE range_parted DETACH PARTITION range_parted_5;
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 550;
                     QUERY PLAN                     
----------------------------------------------------
 Global Index Scan using global_idx on range_parted
   Index Cond: (b = 550)
(2 rows)

SELECT * FROM range_parted WHERE b = 550 LIMIT 5;
 a | b 
---+---
(0 rows)

--Reattach the partition
ALTER TABLE range_parted ATTACH PARTITION range_parted_5 FOR VALUES FROM (400000) to (500000);
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 550;
                     QUERY PLAN                     
----------------------------------------------------
 Global Index Scan using global_idx on range_parted
   Index Cond: (b = 550)
(2 rows)

SELECT * FROM range_parted WHERE b = 550 LIMIT 5;
 a | b 
---+---
(0 rows)

--Drop the partitioned table
DROP TABLE range_parted_5;
EXPLAIN (COSTS OFF) SELECT * FROM range_parted WHERE b = 550;
                     QUERY PLAN                     
----------------------------------------------------
 Global Index Scan using global_idx on range_parted
   Index Cond: (b = 550)
(2 rows)

SELECT * FROM range_parted WHERE b = 550 LIMIT 5;
 a | b 
---+---
(0 rows)

-- Test unique global index
TRUNCATE TABLE range_parted;
INSERT INTO range_parted VALUES(1,2);
INSERT INTO range_parted VALUES(2,2);
CREATE UNIQUE INDEX global_idx_unique ON range_parted(b) global; -- Fail with duplicate
ERROR:  could not create unique index "global_idx_unique"
DETAIL:  Key (b, )=(2, 1) is duplicated.
TRUNCATE TABLE range_parted;
INSERT INTO range_parted VALUES(1,2);
CREATE UNIQUE INDEX global_idx_unique ON range_parted(b) global;
INSERT INTO range_parted VALUES(1,2); -- Fail with duplicate
ERROR:  duplicate key value violates unique constraint "global_idx_unique"
DETAIL:  Key (b, )=(2, 1) already exists.
DROP INDEX global_idx_unique;
INSERT INTO range_parted VALUES(1,2); -- Now this should pass
-- Cleanup
DROP TABLE range_parted;
